### PR TITLE
RobotLog: Add numeric and boolean topics

### PR DESCRIPTION
**Pull request by coderkevin**
This adds double, integer, and boolean topic types

### CODE REVIEW

File: src/main/java/com/systemmeltdown/meltdownlog/topics/DoubleTopic.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+	public void log(double value) {
+		log(value, System.nanoTime());
+	}
+
+	public void log(double value, long nanos) {
+		final double roundedValue = Utils.roundByFactor(value, m_roundingFactor);
+		if (roundedValue == m_lastRoundedValue) {
+			// Don't write the same value over and over again.
+			m_lastRoundedValue = roundedValue;
```
Comment (shadwstalkr) [inline src/main/java/com/systemmeltdown/meltdownlog/topics/DoubleTopic.java]: Line is redundant

File: src/main/java/com/systemmeltdown/meltdownlog/topics/DoubleTopic.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		m_roundingFactor = roundingFactor;
+	}
+
+	public void log(double value) {
+		log(value, System.nanoTime());
+	}
+
+	public void log(double value, long nanos) {
+		final double roundedValue = Utils.roundByFactor(value, m_roundingFactor);
+		if (roundedValue == m_lastRoundedValue) {
```
Comment (shadwstalkr) [inline src/main/java/com/systemmeltdown/meltdownlog/topics/DoubleTopic.java]: Same issue here where the first value might not be written to the log.

File: src/main/java/com/systemmeltdown/meltdownlog/topics/IntegerTopic.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		m_roundingFactor = roundingFactor;
+	}
+
+	public void log(final int value) {
+		log(value, System.nanoTime());
+	}
+
+	public void log(final int value, final long nanos) {
+		final int roundedValue = Utils.roundByFactor(value, m_roundingFactor);
+		if (roundedValue == m_lastRoundedValue) {
```
Comment (shadwstalkr) [inline src/main/java/com/systemmeltdown/meltdownlog/topics/IntegerTopic.java]: Same stuff here

File: src/main/java/com/systemmeltdown/meltdownlog/topics/BooleanTopic.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	}
+
+	public void log(boolean value) {
+		log(value, System.nanoTime());
+	}
+
+	public void log(boolean value, long nanos) {
+		if (value == m_lastValue) {
+			// Don't write the same value over and over again.
+			m_lastValue = value;
```
[the remark this answers is not in the document]
Comment (coderkevin) [reply src/main/java/com/systemmeltdown/meltdownlog/topics/BooleanTopic.java]: Fixed, thanks!

File: src/main/java/com/systemmeltdown/meltdownlog/topics/BooleanTopic.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	public BooleanTopic(String name) {
+		super(name, Boolean.class);
+	}
+
+	public void log(boolean value) {
+		log(value, System.nanoTime());
+	}
+
+	public void log(boolean value, long nanos) {
+		if (value == m_lastValue) {
```
[the remark this answers is not in the document]
Comment (coderkevin) [reply src/main/java/com/systemmeltdown/meltdownlog/topics/BooleanTopic.java]: I have now consolidated this functionality in a single class, DataTopic.
Thanks!

File: src/main/java/com/systemmeltdown/meltdownlog/topics/DoubleTopic.java (hunk 2 above)
Comment (coderkevin) [reply src/main/java/com/systemmeltdown/meltdownlog/topics/DoubleTopic.java]: I have now consolidated this functionality in a single class, DataTopic.
Thanks!

File: src/main/java/com/systemmeltdown/meltdownlog/topics/IntegerTopic.java (hunk 3 above)
Comment (coderkevin) [reply src/main/java/com/systemmeltdown/meltdownlog/topics/IntegerTopic.java]: I have now consolidated this functionality in a single class, DataTopic.
Thanks!

File: src/main/java/com/systemmeltdown/robotlog/topics/DataTopic.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+		if (m_lastDuplicateNanos != Long.MIN_VALUE) {
+			// Finish reporting the duplicate values before reporting this one.
+			writeEntry(m_lastValue, m_lastDuplicateNanos);
+		}
+
+		writeEntry(value, nanos);
+
+		// Now we don't have any unreported values.
+		m_lastDuplicateNanos = Long.MIN_VALUE;
+	}
```
Comment (shadwstalkr) [inline src/main/java/com/systemmeltdown/robotlog/topics/DataTopic.java]: You never set m_lastValue

File: src/test/java/com/systemmeltdown/robotlog/topics/BooleanTopicTest.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+		topic.log(true, nanos + 300);
+		topic.log(false, nanos + 400);
+		topic.log(false, nanos + 500);
+		topic.log(false, nanos + 600);
+		topic.log(true, nanos + 700);
+
+		InOrder inOrder = Mockito.inOrder(writer);
+		inOrder.verify(writer).writeEntry("boolean-topic", false, nanos + 0);
+		inOrder.verify(writer).writeEntry("boolean-topic", true, nanos + 100);
+		inOrder.verify(writer).writeEntry("boolean-topic", false, nanos + 400);
```
Comment (shadwstalkr) [inline src/test/java/com/systemmeltdown/robotlog/topics/BooleanTopicTest.java]: Shouldn't there be a true at nanos + 300?

File: src/main/java/com/systemmeltdown/robotlog/topics/DataTopic.java (hunk 6 above)
Comment (coderkevin) [reply src/main/java/com/systemmeltdown/robotlog/topics/DataTopic.java]: Good catch. Thanks!

File: src/test/java/com/systemmeltdown/robotlog/topics/BooleanTopicTest.java (hunk 7 above)
Comment (coderkevin) [reply src/test/java/com/systemmeltdown/robotlog/topics/BooleanTopicTest.java]: Yes, there should be a true at 300 and a false at 600. Thanks!

Comment (coderkevin) [reply src/test/java/com/systemmeltdown/robotlog/topics/BooleanTopicTest.java]: Fixed in a5701aa

File: src/main/java/com/systemmeltdown/robotlog/topics/DataTopic.java (hunk 6 above)
Comment (coderkevin) [reply src/main/java/com/systemmeltdown/robotlog/topics/DataTopic.java]: Fixed in a5701aa

**Pull request merged**